common/unsafe/src/main/java/org/apache/spark/unsafe/types/UTF8String.java

```diff
@@ -1136,6 +1136,104 @@ public UTF8String replace(UTF8String search, UTF8String replace) {
     return buf.build();
   }
 
+  public UTF8String replace(UTF8String search, UTF8String replace, int collationId) {
+    if (CollationFactory.fetchCollation(collationId).isBinaryCollation) {
+      return this.replace(search, replace);
+    }
+    if (collationId == CollationFactory.LOWERCASE_COLLATION_ID) {
+      return lowercaseReplace(search, replace);
+    }
+    return collatedReplace(search, replace, collationId);
+  }
+
+  public UTF8String lowercaseReplace(UTF8String search, UTF8String replace) {
```

Comment:
is it faster than `collatedReplace`?

Reply:
It is **not** faster. It is a separate method because 
`StringSearch stringSearch = CollationFactory.getStringSearch(this, search, collationId);` fails for `collationId == 1` because _collator_ for that value is _null_ (see `CollationFactory`). For that reason I had to do comparisons the standard way we do it for `UTF8_BINARY_LCASE` - convert everything to lowercase and compare it that way. 

So the idea is to do comparisons on lowercase strings but use the original string for building the final result, to keep all uppercase and lowercase as in the input.

With other type of collations we can do all comparisons on the original input values.

Reply:
Can we create a `CollationFactory` for the collation with id 1?

Reply:
I think @dbatomic knows better than me if we can/should do that.

Reply:
I think there is simply no `com.ibm.icu.text.Collator` instance for **UTF8_BINARY_LCASE**. After all, UTF8String is a Spark concept, and while I suppose that we could introduce some kind of wrapper collator for UTF8_BINARY_LCASE, it wouldn't seem like a good idea to me

Instead, I think there should be an easy way to get a `com.ibm.icu.text.StringSearch` instance without passing a collationId (CollationFactory currently doesn't support this, but @miland-db can modify it in a way that would allow optional collationId: `new StringSearch(pattern, target)`)

After these changes, I believe `lowercaseReplace` and `collatedReplace` can essentially be combined into a single function, where for **UTF8_BINARY_LCASE** it should be enough to pass `.toLowerCase()` for both **this** & **search** parameters (with no collationId)

Reply:
I managed to refactor this part of the code and remove `lowercaseReplace`. Now we have getStringSearch method that works with `UTF8_BINARY_LCASE` collation.